### PR TITLE
Add a k-d tree implementation, and integrate it with Hdbscan and KNN.

### PR DESCRIPTION
**Pull request by geoffreydstewart**
### Description
This PR introduces a k-d tree implementation, which can be used for nearest neighbour queries, and integrates this feature with Hdbscan and KNN. Initial performance measurements demonstrate that this k-d tree implementation is, in general, much faster than using a brute-force approach for larger datasets.

After these changes have been reviewed, there is a required change to the Hdbscan tutorial that will be made.

This PR supersedes #230 which contains a lot of interesting code review discussion, but was closed as the result of the branch being recreated. 

### Motivation
K-d trees can provide faster nearest neighbour queries compared to a brute-force technique. This results in faster model training times.

### Paper reference
This is the original paper which proposes a k-d tree:
J.L. Bentley "Multidimensional Binary Search Trees Used for Associative Searching", Commun. ACM, Vol 18, Sept. 1975, 509–517

### CODE REVIEW

File: Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+        }
+
+        // points[lowIndex] < points[midIndex] < points[right]
+        return midIndex;
+    }
+
+    /**
+     * Partition the array into two parts around the provided array index. Those less than the value at the provided
+     * array index will be on the left. Those greater than or equal to the value at the provided array index will fall
+     * on the left. The operation is performed for a specific dimension of the vectors defined by the comparator.
```
Comment (Craigacp) [inline Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java:250]: I think this should say "on the right" as the values less than are on the left.

File: Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+     * @return The correct, updated index of the pivot point.
+     */
+    private static int partitionOnIndex(IntAndVector[] points, int left, int right, int pivotIndex, int dimension) {
+        IntAndVector pivot = points[pivotIndex];
+        // for now, move the pivot to the end of the array
+        swap(points, right, pivotIndex);
+
+        int store = left;
+        for (int idx = left; idx < right; idx++) {
+            if (compareByDimension(points[idx], pivot, dimension) <= 0) {
```
Comment (Craigacp) [inline Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java:266]: Shouldn't this be a strict less than? The docs say it's less than on the left and greater than or equal on the right.

File: Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+        // Generate the below/left and above/right subtrees.
+        medianNode.setBelow(generateTree(d, maxD, points, left, left+median-2, distanceType));
+        medianNode.setAbove(generateTree(d, maxD, points, left+median, right, distanceType));
+        return medianNode;
+    }
+
+    /**
+     * Set the median point for an array of {@link IntAndVector}s based, for a specific dimension, through recursive partitioning
+     * ensuring that points before it (with lower index) will be <= median, although not sorted, and points after it
+     * (with higher index) will be >= median, again not sorted. The order of the array will almost certainly be changed.
```
Comment (Craigacp) [inline Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java:184]: So this says that values equal to the median can be found on both sides of the tree, but that seems to conflict with the documented behaviour of the `partitionOnIndex` function (though not its actual behaviour).

File: Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+            try {
+                boolean finished = executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.MINUTES);
+                if (!finished) {
+                    throw new RuntimeException("Parallel execution failed");
+                }
+            } catch (InterruptedException e) {
+                throw new RuntimeException("Parallel execution failed", e);
+            }
+        }
+        return new ArrayList<>(Arrays.asList(indexDistancePairListArray));
```
Comment (Craigacp) [inline Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java:132]: I think you missed converting this over to just returning `Arrays.asList`.

**Review by Craigacp — CHANGES_REQUESTED** [review]
I'd like to properly nail down the semantics of the k-d tree building (which may just be documentation fixes now), and there are a couple of other small points.

File: Clustering/Hdbscan/src/test/java/org/tribuo/clustering/hdbscan/TestHdbscan.java [hunk 5]
```diff
@@ -93,7 +94,7 @@ public void testInvocationCounter() {
         CSVDataSource<ClusterID> csvSource = new CSVDataSource<>(Paths.get("src/test/resources/basic-gaussians.csv"),rowProcessor,false);
         Dataset<ClusterID> dataset = new MutableDataset<>(csvSource);
 
-        HdbscanTrainer trainer = new HdbscanTrainer(7, DistanceType.L2, 7,4);
+        HdbscanTrainer trainer = new HdbscanTrainer(7, DistanceType.L2, 7,4, NeighboursQueryFactoryType.BRUTE_FORCE);
```
[the remark this answers is not in the document]
Comment (geoffreydstewart) [reply Clustering/Hdbscan/src/test/java/org/tribuo/clustering/hdbscan/TestHdbscan.java:99]: Yes, it's good to demonstrate the programatic execution path. For the tutorial update, II think 'll use an enum constant.

File: Math/src/main/java/org/tribuo/math/neighbour/NeighboursQueryFactory.java [hunk 6]
```diff
@@ -20,19 +20,28 @@
...
      * Constructs a nearest neighbour query object using the supplied array of {@link SGDVector}.
      * @param data An array of {@link SGDVector}.
      */
     NeighboursQuery createNeighboursQuery(SGDVector[] data);
 
+    /**
+     * Gets the {@link DistanceType} set on this object.
+     * @return The distance function.
+     */
+    DistanceType getDistanceType();
```
[the remark this answers is not in the document]
Comment (geoffreydstewart) [reply Math/src/main/java/org/tribuo/math/neighbour/NeighboursQueryFactory.java:43]: Sure, let's avoid that breaking change later.

File: Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTreeFactory.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Constructs a k-d tree nearest neighbor query factory object using the supplied parameters.
+     * @param distanceType The distance function.
+     * @param numThreads The number of threads to be used to parallelize the computation.
+     */
+    public KDTreeFactory(DistanceType distanceType, int numThreads) {
+        this.distanceType = distanceType;
+        this.numThreads = numThreads;
```
[the remark this answers is not in the document]
Comment (geoffreydstewart) [reply Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTreeFactory.java:50]: Yep, added the `postConfig`s and quick corresponding tests for bad config case.

File: Clustering/Hdbscan/src/test/java/org/tribuo/clustering/hdbscan/TestHdbscan.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply Clustering/Hdbscan/src/test/java/org/tribuo/clustering/hdbscan/TestHdbscan.java:99]: Sure, that's fine. I've hit issues where different constructors cause different failures so I like to exercise each base constructor when testing now.

**Review by Craigacp — APPROVED** [review]
LGTM

**Pull request merged**